src/snap/proxy/stream_collector_proxy.cc

```diff
@@ -0,0 +1,301 @@
+                                this, task_id, context, stream);
+        auto recvch = std::async(std::launch::async, &StreamCollectorImpl::streamRecv,
+                                this, task_id, context, stream);
+        auto errch = std::async(std::launch::async, &StreamCollectorImpl::errorSend,
+                                this, context, stream);
+
+        auto do_puts = std::async(std::launch::async, &StreamCollectorImpl::PutSendMetsAndErrMsg,
+                                this, context);
+
+        _stream_collector->stream_metrics();
+
+        return Status::OK;
+    } catch(PluginException &e) {
+        return Status(StatusCode::UNKNOWN, e.what());
+    }
+}
+
+bool StreamCollectorImpl::PutSendMetsAndErrMsg(ServerContext* context) {
+    std::vector<Metric> recv_mets, send_mets;
+    while(!context->IsCancelled()) {
+        if (_stream_collector->put_mets()) {
+            _sendChan.put(_stream_collector->put_metrics_out());
+            _stream_collector->set_put_mets(false);
+        }
+        if (_stream_collector->put_err()) {
+            _errChan.put(_stream_collector->put_err_msg());
+            _stream_collector->set_put_err(false);
+        }
+        if (_recvChan.get(recv_mets, false)) {
+            _stream_collector->get_metrics_in(recv_mets);
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    }
+    _stream_collector->set_context_cancelled(true);
+}
+
+bool StreamCollectorImpl::errorSend(ServerContext* context,
+                                    ServerReaderWriter<CollectReply, CollectArg>* stream) {                            
+    try {
+        while (!context->IsCancelled()) {
+            std::string err;
+            if (_errChan.get(err)) {
+                _err_reply.set_error(err);
+                stream->Write(_collect_reply);
+                err.clear();
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        }
+        _errChan.close();
+        _stream_collector->set_context_cancelled(true);
+        return true;
+    } catch (PluginException &e) {
+        std::cout << "Error" << std::endl;
+        return false;
+    }
+}
+
+bool StreamCollectorImpl::metricSend(const std::string &taskID,
+                                    ServerContext* context,
+                                    ServerReaderWriter<CollectReply, CollectArg>* stream) {
+    try {
+        std::chrono::system_clock::time_point start = std::chrono::system_clock::now();
+        while (!context->IsCancelled()) {
+            std::vector<Metric> send_mets;
+            if (_sendChan.get(send_mets)) {
+                if (!send_mets.empty()) {
+                    for (Metric met : send_mets) {
+                        *_metrics_reply.add_metrics() = *met.get_rpc_metric_ptr();
+                        if ((_metrics_reply.metrics_size() >= _max_metrics_buffer) && 
```

Comment:
this condition is not proper. metrics_size should be equal to max_metrics_buffer instead of grater or equal